lib/c.c

```diff
@@ -529,11 +529,16 @@ int fputc(int c, FILE *stream)
 /* Non-portable: Assume page size is 4KiB */
 #define PAGESIZE 4096
 
+#define CHUNK_SIZE_FREED_MASK 1
+#define CHUNK_SIZE_SZ_MASK 0xFFFFFFFE
+#define CHUNK_GET_SIZE(size) (size & CHUNK_SIZE_SZ_MASK)
+#define IS_CHUNK_GET_FREED(size) (size & CHUNK_SIZE_FREED_MASK)
+#define CHUNK_SET_FREED(size) (size | CHUNK_SIZE_FREED_MASK)
```

Comment:
How about rewriting `CHUNK_SET_FREED` as following?
```c
#define CHUNK_SET_FREED(size) \
    size |= CHUNK_SIZE_FREED_MASK
```

Reply:
I encounter some trouble when modifying the macro.

Then, I use the following code to test a function-like macro:
```c
/* test.c */
#define MACRO(a) a += 10

int main()
{
    int x = 10;
    printf("%d\n", x);
    MACRO(x);
    printf("%d\n", x);

    return 0;
}
```
```
$ qemu-arm out/shecc-stage2.elf -o test test.c
Error Unrecognized statement token at source location 16802
#define MACRO(a) a += 10
                   ^ Error occurs here
Abnormal program termination
```
This implies that shecc cannot parse the macro containing compound operator(s) correctly.

Reply:
The current parser lacks of the ability to deal with compound assignment operators. So, you have to change the statements.

Reply:
Because the current parser also lacks of the ability to handle assignment operator. Therefore, I think I will use two functions called `chunk_set_freed()` and `chunk_clear_freed()` to perform the operations instead of using the original macros.

```c
void chunk_set_freed(chunk_t *chunk) 
{
        chunk->size |= CHUNK_SIZE_FREED_MASK;
}

void chunk_clear_freed(chunk_t *chunk) 
{
        chunk->size &= CHUNK_SIZE_SZ_MASK;
}
```

Reply:
> I think I will use two functions called `chunk_set_freed()` and `chunk_clear_freed()` to perform the operations instead of using the original macros.

It looks good to me.

Reply:
> the current parser also lacks of the ability to handle assignment operator.

@ChAoSUnItY, The compound assignment operators consist of a binary operator and the simple assignment operator. At present, there is no support for compound assignment operators. Can you evaluate the feasibility to consolidate the parser for compound assignment operators?

Reply:
Here are the examples that shecc cannot deal with.
```c
#define MACRO1(variable, val) \
               variable = variable + val + 10
#define MACRO2(variable, val) \
               variable += val + 10
```
These macros contain an assignment operator and a compound assignment operator, and shecc cannot parse them correctly.

Reply:
> Here are the examples that shecc cannot deal with.

Create an issue for tracking.

Reply:
> The compound assignment operators consist of a binary operator and the simple assignment operator. At present, there is no support for compound assignment operators. 

The current situation is that `read_body_assignment` is failed to realize the macro variable and thus returns false since it's possibly not a local variable nor a global variable, which ultimately results in the parser failed to parse current variable and terminated at the operator.
 
> Can you evaluate the feasibility to consolidate the parser for compound assignment operators?

Currently, it seems rather tricky to implement due to the technical debt of macro parsing design (we're not able to determine which syntax parsing we should use to parse macro parameter). Moreover, if we predicts it's a lvalue (which in this case it should be) and we use `read_lvalue` function to parse the parameter, the function requires a non-null `var_t`, which is impossible because we haven't even parse macro parameter's actual expression; if we predicts it's an expression, we'll lose the target address of the lvalue.

This probably could be resolved after we implement [tangle](https://github.com/ChAoSUnItY/tangle) back to shecc since macro expansion would only happens in lexer and parser doesn't have to predict what parsing strategy should be used for the expanded macro parameter.